flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaRecordSerializationSchemaBuilder.java

```diff
@@ -0,0 +1,313 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.kafka.sink;
+
+import org.apache.flink.api.common.serialization.SerializationSchema;
+import org.apache.flink.streaming.connectors.kafka.partitioner.FlinkKafkaPartitioner;
+
+import org.apache.flink.shaded.guava30.com.google.common.cache.CacheBuilder;
+import org.apache.flink.shaded.guava30.com.google.common.cache.CacheLoader;
+import org.apache.flink.shaded.guava30.com.google.common.cache.LoadingCache;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.serialization.Serializer;
+
+import javax.annotation.Nullable;
+
+import java.util.Map;
+import java.util.OptionalInt;
+import java.util.function.Function;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Builder to construct {@link KafkaRecordSerializationSchema}.
+ *
+ * <p>This class should give a first entrypoint when trying to serialize elements to {@link
+ * ProducerRecord}. The following examples show some of the possibilities.
+ *
+ * <pre>Simple key-value serialization:
+ * {@code
+ * KafkaRecordSerializationSchema.builder("topic")
+ *     .withKeySerializationSchema(new SimpleStringSchema())
```

Comment:
We are actually mixing two patterns here: Wither pattern and builder pattern. There seem to be a few instances in the code-base so it might not be critical.
In Buidler, you usually have `X.builder().setA().setB().build()`. `X.builder()` returns a class `XBuilder` that is a POJO and creates an immutable object. An example is `AbstractFileSourceBuilder`.
In Wither, you have `X.factory()/new X().withA().withB()`. Where there is only X from the user's perspective that is usually immutable and returns a new instance with modified properties on `withA`. A good example is `WatermarkStrategy`.

A hybrid naming is used for example in `RowFormatBuilder`. So `FileSource` and `FileSink` are implemented asymmetrically.